From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Mon, 15 May 2023 19:00:02 -0600
Subject: [PATCH] wined3d: Avoid sysmem pinning streaming buffer.

CW-Bug-Id: #22259

Currently that happens on _SOFTWARE_VERTEXPROCESSING devices and
prevents _MAP_DISCARD and _MAP_NOOVERWRITE optimization.
Valve-commit: 9754a5cf40b1fee0a45ee7ea1c58802486478a96
Picked: clean
Class: valve
Evidence: docs/evidence/2026-09-28-wine-proton-rebase.md
Offered-upstream: no
---
 dlls/wined3d/buffer.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index 5e1f0de..24949a7 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -971,7 +971,7 @@ static HRESULT wined3d_buffer_init(struct wined3d_buffer *buffer, struct wined3d
         access |= WINED3D_RESOURCE_ACCESS_MAP_W;
 
     if (FAILED(hr = resource_init(resource, device, WINED3D_RTYPE_BUFFER, format,
-            WINED3D_MULTISAMPLE_NONE, 0, desc->usage, desc->bind_flags, access,
+            WINED3D_MULTISAMPLE_NONE, 0, desc->usage & ~WINED3DUSAGE_CS, desc->bind_flags, access,
             desc->byte_width, 1, 1, desc->byte_width, parent, parent_ops, &buffer_resource_ops)))
     {
         WARN("Failed to initialize resource, hr %#lx.\n", hr);
@@ -984,8 +984,8 @@ static HRESULT wined3d_buffer_init(struct wined3d_buffer *buffer, struct wined3d
     TRACE("buffer %p, size %#x, usage %#x, memory @ %p.\n",
             buffer, buffer->resource.size, buffer->resource.usage, buffer->resource.heap_memory);
 
-    if (device->create_parms.flags & WINED3DCREATE_SOFTWARE_VERTEXPROCESSING
-            || (desc->usage & WINED3DUSAGE_MANAGED))
+    if (!(desc->usage & WINED3DUSAGE_CS) && (device->create_parms.flags & WINED3DCREATE_SOFTWARE_VERTEXPROCESSING
+            || (desc->usage & WINED3DUSAGE_MANAGED)))
     {
         /* SWvp and managed buffers always return the same pointer in buffer
          * maps and retain data in DISCARD maps. Keep a system memory copy of
@@ -1364,7 +1364,7 @@ static HRESULT wined3d_streaming_buffer_prepare(struct wined3d_device *device,
     TRACE("Growing buffer to %u bytes.\n", size);
 
     desc.byte_width = size;
-    desc.usage = WINED3DUSAGE_DYNAMIC;
+    desc.usage = WINED3DUSAGE_DYNAMIC | WINED3DUSAGE_CS;
     desc.bind_flags = buffer->bind_flags;
     desc.access = WINED3D_RESOURCE_ACCESS_GPU | WINED3D_RESOURCE_ACCESS_MAP_W;
     desc.misc_flags = 0;
